source/grid/grid_in.cc

```diff
@@ -2820,6 +2824,529 @@ GridIn<dim, spacedim>::read_assimp(const std::string &filename,
+                  return exodus_to_deal[face_n];
+                }
+              default:
+                Assert(false, ExcNotImplemented());
+            }
+          break;
+        case 3:
+          switch (type)
+            {
+              case ReferenceCell::Type::Tet:
+                {
+                  AssertIndexRange(face_n, 4);
+                  constexpr std::array<unsigned int, 4> exodus_to_deal{
+                    {1, 3, 2, 0}};
+                  return exodus_to_deal[face_n];
+                }
+              case ReferenceCell::Type::Hex:
+                {
+                  AssertIndexRange(face_n, 6);
+                  constexpr std::array<unsigned int, 6> exodus_to_deal{
+                    {2, 1, 3, 0, 4, 5}};
+                  return exodus_to_deal[face_n];
+                }
+              default:
+                Assert(false, ExcNotImplemented());
+            }
+          break;
+        default:
+          break;
+      }
+    Assert(false, ExcInternalError());
+    return numbers::invalid_unsigned_int;
+  }
+
+  class ExcExodusII : public ExceptionBase
+  {
+  public:
+    ExcExodusII(const int error_code)
+      : error_code(error_code)
+    {}
+
+    virtual void
+    print_info(std::ostream &out) const override
+    {
+      out << "Error code is " << error_code << '\n';
+      out << "String description: " << ex_strerror(error_code) << '\n';
+    }
+
+    const int error_code;
+  };
+
+#  define AssertThrowExodusII(error_code) \
+    AssertThrow(error_code == EX_NOERR, ExcExodusII(error_code));
+
+  // Associate deal.II boundary ids with sidesets (a face can be in multiple
+  // sidesets - to translate we assign each set of side set ids to a
+  // boundary_id)
+  template <int dim, int spacedim = dim>
+  std::pair<SubCellData, std::vector<std::vector<int>>>
+  read_exodus_sidesets(const int                         ex_id,
+                       const int                         n_side_sets,
+                       const std::vector<CellData<dim>> &cells)
+  {
+    SubCellData                   subcelldata;
+    std::vector<std::vector<int>> boundary_id_to_sideset_ids;
+    // boundary id 0 is the default
+    boundary_id_to_sideset_ids.emplace_back();
+    // Exodus prints warnings if we try to get empty sets so always check first
+    if (n_side_sets > 0)
+      {
+        std::vector<int> side_set_ids(n_side_sets);
+        int ierr = ex_get_ids(ex_id, EX_SIDE_SET, side_set_ids.data());
+        AssertThrowExodusII(ierr);
+
+        // First collect all side sets on all boundary faces (crudely indexed
+        // here as max_faces_per_cell * cell_n + face_n). We then sort and
+        // uniquify the side sets so that we can convert a set of side set
+        // indices into a single deal.II boundary id (and save the
+        // correspondence).
+        constexpr auto max_faces_per_cell = GeometryInfo<dim>::faces_per_cell;
+        std::vector<std::vector<int>> face_side_sets(max_faces_per_cell *
```

Comment:
That is a lot of memory. What stops you from filling boundary_id_to_sideset_ids directly as you go?

Reply:
My intent with the current approach is that we record all sidesets and then uniquify that list to get `boundary_id_to_sideset_ids`. I'm not sure how to do it incrementally since we won't know which elements have the same sidesets until we have read everything. There might be a fix - I'll think about it.

Reply:
I switched to using `std::map<int, std::vector<int>>` since the majority of these entries will be empty.

Reply:
I am pretty sure its not possible to build that output array incrementally - I can write down a detailed example if you want to see it.

Reply:
how about:
```
std::map m;  // sideset |-> index

for each cell:
for each face f at boundary:
 1. grab list of sidesets X for face f
 2. put X into canonical form (sort?)
 3. If X is in m, assign its value m[X], if X is not in m, insert it with the next free boundary id.
```

Reply:
That's not going to work because we have to loop over the side set numbers (i.e., we read one side set at a time) and in each side set number we have the lists of element and face numbers. Hence we cannot establish canonical values until we have read everything.

Reply:
Ah, okay. That is annoying. Anyways, the map is much better than the gigantic vector (unless this also happens for internal faces?)

Reply:
ExodusII doesn't store any topological information so it is possible to assign side sets to internal face ids (though the documentation states that a side should be uniquely identified by a face number and element number, so I think this officially not supported).